Add experimental rematerialization decorator

mattjj opened the pull request:
## The problem

We want to allow users to control how reverse-mode autodiff saves values from the forward pass. In particular, we want it to be easy to signal that a function shouldn't have any of its intermediate residuals stored for the backward pass, and instead those values should be recomputed from the function's saved inputs. This allows users greater control over the time/memory tradeoffs of their computations. (This feature is especially handy for accelerators on which memory access is much more expensive than FLOPs are.)

In JAX terms, since we implement reverse-mode as a composition of forward-mode, partial evaluation, and transposition, we want users to control how partial evaluation behaves.

This PR adds a `checkpoint` decorator (alias: `remat`) for this purpose.

Fixes #1732.

## API examples

### Example 1: grad-of-jit

Consider this grad-of-jit situation:

```python
import jax.numpy as np
from jax import jit, grad, remat

@remat
def g(x):
  return np.sin(np.sin(x))

@jit
def f(x):
  return g(x)

grad(f)(2.0)
```

If we run this _without_ the remat decorator, here are the jaxprs we end up lowering to XLA, one for the forward pass and one for the backward pass (by adding a `print(jaxpr)` to `xla._xla_callable`):

```
{ lambda  ;  ; a b.
  let c = sin a
      d = sin c
      e = cos a
      f = cos c
  in [d, *, e, f] }

{ lambda  ;  ; a b c.
  let d = mul c b
      e = mul d a
  in [e] }
```

Here's what we see _with_ the `remat` decorator:

```
{ lambda  ;  ; a b.
  let c = sin a
      d = sin c
  in [d, *, a] }

{ lambda  ;  ; a b.
  let c = remat_call a b
        { lambda  ;  ; a b.
          let c = sin a
              d = cos c
              e = mul b d
              f = cos a
              g = mul e f
          in [g] } [  ;  ]
  in [c] }
```

Notice how there are no residuals passed into the backward pass as constants, and we're computing `sin(2.0)` both in the forward pass and in the backward pass.


### Example 2: jit-of-grad

Here's a jit-of-grad situation:

```python
import jax.numpy as np
from jax import jit, value_and_grad, remat

@remat
def g(x):
  return np.sin(np.sin(x))

jit(grad(g))(2.0)
```

```
{ lambda  ;  ; a.
  let b = sin a
      c = sin b
      d = remat_call a 1.0
        { lambda  ;  ; a b.
          let c = sin a
              d = cos c
              e = mul b d
              f = cos a
              g = mul e f
          in [g] } [  ;  ]
  in [c, d] }
```

Again we're computing `sin(2.0)` twice. Not shown is the fact that the translation rule for lowering `remat_call` to XLA HLO includes some HLO widgets that will foil XLA's CSE optimization across `remat_call` boundaries, even though it's all being lowered to one computation here.

### Example 3: differentiating Python control flow under `remat`

```python
import jax.numpy as np
from jax import jit, grad, linearize, remat, make_jaxpr

@partial(remat, concrete=True)
def g(x):
  if x > 0.:
    return np.sin(np.sin(x))
  else:
    return np.cos(np.cos(x))

def f(x):
  x = 3. * x
  x = g(x)
  x = 4. * x
  return x

print(grad(f)(2.))
jaxpr = make_jaxpr(linearize(f, 2.)[1])(1.)
print(jaxpr)
```

```
11.075182

{ lambda  ;  ; a.
  let b = mul a 3.0
      c d = remat_call 6.0 b
        { lambda  ;  ; a b.
          let c = sin a
              d = sin c
              e = cos a
              f = mul b e
              g = cos c
              h = mul f g
          in [d, h] } [  ;  ]
      e = mul d 4.0
  in [e] }
```

Functions with `remat` can still support differentiating through Python control flow! At least, when you pass `concrete=True`. There's a tradeoff here: to handle Python control flow, in some cases involving `jit` we might end up doing redundant FLOPs, as in the [Three Sines Problem described below](https://github.com/google/jax/pull/1749#issuecomment-558267584), so we wanted to make this feature opt-in.

Notice that in this last jaxpr, the `remat_call` is applied to an argument that is a literal `6.0`. We've prevented that from being partially evaluated through the called jaxpr on the JAX side, and moreover the lowering of `remat_call` will prevent XLA from doing that constant folding as well.

### Example 4: scanning a `remat` function

```python
import jax.numpy as np
from jax import lax
from jax import vjp, remat, make_jaxpr

def f(c, x):
  return np.sin(c), None

def foo(x):
  y, _ = lax.scan(remat(f), x, np.arange(3.))
  return y

_, foo_vjp = vjp(foo, 4.)
foo_vjp(1.)  # added a `print(jaxpr)` in ad.backward_pass
```

```
{ lambda c d ;  ; a b.
  let e f = scan[ forward=True
                  length=3
                  jaxpr={ lambda  ;  ; a b c d e.
                          let f g = remat_call d e b
                                { lambda  ;  ; a b c.
                                  let d = sin a
                                      e = cos a
                                      f = mul c e
                                  in [d, f] } [  ;  ]
                          in [*, g] }
                  num_consts=0
                  num_carry=2
                  linear=[True, True, True, False, False] ] * b * c d
  in [*, f] }
```

Notice the `sin` and `cos` computation in the  scanned function! Compare to not using the `remat` decorator:

```
{ lambda c ;  ; a b.
  let d e = scan[ forward=True
                  length=3
                  jaxpr={ lambda  ;  ; a b c d.
                          let e = mul b d
                          in [*, e] }
                  num_consts=0
                  num_carry=2
                  linear=[True, True, True, False] ] * b * c
  in [*, e] }
```

### Example 5: recursive checkpointing

```python
def recursive_checkpoint(funs):
  if len(funs) == 1:
    return funs[0]
  elif len(funs) == 2:
    f1, f2 = funs
    return lambda x: f1(f2(x))
  else:
    f1 = recursive_checkpoint(funs[:len(funs)//2])
    f2 = recursive_checkpoint(funs[len(funs)//2:])
    return lambda x: f1(remat(f2)(x))

# forward pass
f = recursive_checkpoint([np.cos, np.sin, np.cos, np.sin])
print(make_jaxpr(f)(4.))

# forward and backward pass
f = recursive_checkpoint([np.sin, np.sin, np.sin, np.sin])
print(make_jaxpr(value_and_grad(f))(4.))

# longer forward pass
f = recursive_checkpoint([np.cos, np.sin, np.cos, np.sin,
                          np.cos, np.sin, np.cos, np.sin])
print(make_jaxpr(f)(4.))
```

```
# forward pass
{ lambda  ;  ; a.
  let b = remat_call[ concrete=False ] a
        { lambda  ;  ; a.
          let b = sin a
              c = cos b
          in [c] } [  ;  ]
      c = sin b
      d = cos c
  in [d] }

# forward and backward pass
{ lambda  ;  ; a.
  let b = sin a
      c = sin b
      d = sin c
      e = sin d
      f = cos d
      g = mul 1.0 f
      h = cos c
      i = mul g h
      j = remat_call[ concrete=False ] a i
        { lambda  ;  ; a b.
          let c = sin a
              d = cos c
              e = mul b d
              f = cos a
              g = mul e f
          in [g] } [  ;  ]
  in [e, j] }

# longer forward pass
{ lambda  ;  ; a.
  let b = remat_call[ concrete=False ] a
        { lambda  ;  ; a.
          let b = remat_call[ concrete=False ] a
                { lambda  ;  ; a.
                  let b = sin a
                      c = cos b
                  in [c] } [  ;  ]
              c = sin b
              d = cos c
          in [d] } [  ;  ]
      c = remat_call[ concrete=False ] b
        { lambda  ;  ; a.
          let b = sin a
              c = cos b
          in [c] } [  ;  ]
      d = sin c
      e = cos d
  in [e] }
```

## Implementation

The basic design here is to create a new call primitive, like `core.call_p` (which doesn't _necessarily_ raise the abstraction level of its inputs), that has a special partial evaluation rule: stage out the full computation, not just the parts that couldn't be partially evaluated. We call the new primitive `remat_call_p`.

The implementation of the partial evaluation rule has three main steps:

  1. trace the full jaxpr of the called function, treating all inputs as "unknown" (but potentially using concrete avals for the arguments which are actually "known", if `concrete=True`),
  1. process the full jaxpr to determine which outputs are meant to be "known"/"unknown", and to prune any extra computation,
  1. compute any "known" values we need (as an optimization, extract values from any concrete avals instead of re-evaluating them).

Otherwise, it looks a lot like the standard call partial evaluation rule, which is `JaxprTrace.process_call`.

To make `remat_call_p` transposable, we need to upgrade our `ad.backward_pass` code to do some nonlinear evaluation (to rematerialize the residuals needed).

To translate `remat_call_p` to XLA in a way that prevents XLA's CSE optimization from foiling our rematerialization plans, we use a special widget that (according to friendly XLA team members) will in effect force a false data dependence to avoid CSE.

This change also adjusts our abstract evaluation rules for concrete values not to promote to the shaped level. (It was a long time ago, but I think we had that policy only because we didn't have a use case for performing FLOPs while creating jaxprs.) To ensure that this change doesn't accidentally lead to more FLOPs in user code, I tried running the test suite with an `assert False` in the new concrete-evaluation path and checked that it was never hit except in the new `remat` tests.

The upgrade to ad.backward_pass was co-authored with @dougalm, and is essentially part of #1719.

mattjj committed:
Add experimental rematerialization decorator

We want to allow users to control how reverse-mode autodiff saves values
from the forward pass. In particular, we want it to be easy to signal
that a function shouldn't have any of its intermediate residuals stored
for the backward pass, and instead those values should be recomputed
from the function's saved inputs. (This feature is especially handy for
accelerators on which memory access is much more expensive than FLOPs
are.) In JAX terms, since we implement reverse-mode as a composition of
forward-mode, partial evaluation, and transposition, we want users to
control how partial evaluation behaves.

See https://github.com/google/jax/pull/1749 for more.

Co-authored-by: Dougal Maclaurin <dougalm@google.com>

mattjj commented:
@dougalm spotted an issue: when tracing on concrete values (to enable Python control flow) we might still incur extra work if we have `jit` mixed in in the right way. Concretely, here's the "three sines" problem he came up with:

```python
from jax import jit, grad, remat
import jax.numpy as np

@jit
def g(x):
  return f(2., x)

@remat
def f(x, y):
  return np.sin(x) * y

grad(g)(3.)  # sin_p.impl evaluated three times!
```

Running this, `sin(2.)` is evaluated _three_ times rather than the intended two. The issue is that the residual value `sin(2.0)` is computed but thrown away when tracing the function on concrete avals (i.e. in the first of the three steps in the OP, it's thrown away when `mul` has one concrete and one shape-abstracted-by-`jit` argument), then recomputed when we need to evaluate the known outputs (i.e. in the third of the three steps in the OP). I believe `jit` is essential here (and that's why JAX is harder than Autograd!).

The fix we want to try is to make tracing on concrete values (to support Python control flow) optional, at the risk of doing redundant FLOPs when `jit` is involved in slightly tricky ways.

jekbradbury commented:
If `g` is jitted, that's an assertion by the user that it can be faithfully replaced with a shape-abstracted version of itself--I would be surprised that anything inside `g` (or anything called by anything inside `g`) would run with concrete avals?

mattjj committed:
try remat_call partial-eval into two remat_calls

The idea here was for the resulting jaxpr to have a purely nonlinear
remat_call and a linear one with no primals to evaluate. (I wanted to
avoid having to recurse into all calls in _eval_primal in
backward_pass.) But the issue is that makes jaxprs not round-trippable,
since the first remat_call, depending only on constants, would get
partial-eval'd away at the first attempted round-trip. And we round-trip
in partial_eval_jaxpr, particularly for partial eval of scan. That meant
remat of scan didn't work, and that's no good!

mattjj committed:
make remat_call partial-eval into one remat_call

mattjj commented:
I learned something interesting! In 9987f73 I tried to make the partial eval rule of `remat_call` stage out _two_ `remat_call`s, which in the context of reverse-mode ad would have the first only involve nonlinear computations (not depending on the tangents) and output residuals, while the second was linear. The aim was to be cleaner conceptually (don't mix linear and nonlinear in a single call), and also to avoid unnecessary recursing into calls when evaluating nonlinear constants in `ad.backward_pass`.

But that ran into an interesting problem: it would form jaxprs like this:

```
{ lambda  ;  ; a b.
  let c d = remat_call[ concrete=False ] 2.0 *
        { lambda  ;  ; a b.
          let c = sin a
              e = cos a
              f = cos c
          in [e, f] } [  ;  ]
      e f = remat_call[ concrete=False ] * b c d
        { lambda  ;  ; a b c e.
          let d = mul b c
              f = mul d e
          in [*, f] } [  ;  ]
  in [*, f] }
```

and that is not "round-trippable" in the sense that if we do a `make_jaxpr` on an `eval_jaxpr` the first `remat_call` will get partial-evaluated away (it has no data dependence on the input arguments). We use round-tripping like that in `partial_eval_jaxpr` e.g. to implement partial eval of `scan`. That meant that by trying to stage out separate nonlinear and linear `remat_call`s, I broke `scan(remat(f), ...)`.

So I reverted to staging out a single (mixed linear/nonlinear) `remat_call`, and just special-casing it in `ad.backward_pass` to avoid unnecessary recursions into other call primitives (which are always linearized by our partial-eval-of-jvp machinery).

mattjj committed:
raise error if we do concrete aval FLOPs w/o remat

mattjj commented:
cc @joschu

joschu commented:
very cool!

mattjj commented:
Finally documented this feature in #13756.
